In [16]:
import pandas as pd
from statsbombpy import sb
import numpy as np
pd.set_option('display.max_columns', None)

# Femenil

In [17]:
women = pd.read_csv('wc_wom_2023.csv')

In [18]:
women.head()

Rk            Player Pos               Squad   Age    Born  90s  CrdY  \
0   1  Teresa Abelleira  MF            es Spain  23.0  2000.0  6.7     0   
1   2   Deborah Abiodun  MF          ng Nigeria  19.0  2003.0  1.0     0   
2   3           Adriana  MF           br Brazil  26.0  1996.0  2.9     0   
3   4          Lily Agg  MF  ie Rep. of Ireland  29.0  1993.0  1.2     0   
4   5   Eseosa Aigbogun  DF      ch Switzerland  30.0  1993.0  3.5     0   

   CrdR  2CrdY  Fls  Fld  Off  Crs  Int  TklW  PKwon  PKcon  OG  Recov  Won  \
0     0      0    9   10    0   55    9    10      0      0   0     70   28   
1     1      0    2    0    0    0    1     5      0      0   0     11    1   
2     0      0    0    4    0    7    1     3      0      0   0     21    4   
3     0      0    4    3    0    0    1     1      0      0   0     11    3   
4     0      0    4    1    1    8    1     7      0      0   0     30    2   

   Lost  Won%  Matches  
0    14  66.7  Matches  
1     2  33.3  Matches  
2     2  66.7  Matches  
3     4  42.9  Matches  
4     6  25.0  Matches

In [19]:
mids = women[women['Pos'].str.contains("MF")]
mids = mids[(women['Age'] >= 18) & (women['Age'] < 29)]

C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\2516747602.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  mids = mids[(women['Age'] >= 18) & (women['Age'] < 29)]


- Intercepciones (Int): Un buen número de intercepciones refleja que el jugador está cortando pases y controlando el medio campo defensivamente.
- Entradas ganadas (TklW): Los jugadores de contención generalmente tienen un alto número de entradas exitosas.
- Faltas cometidas (Fls): Los centrocampistas defensivos suelen cometer más faltas debido a la naturaleza de su trabajo, que implica frenar el avance de los rivales.
- Recuperaciones (Recov): Número de balones sueltos que el jugador recupera.
- Duelos aéreos ganados (Won%): Muchos centrocampistas defensivos se involucran en duelos aéreos para recuperar balones en el mediocampo.

In [20]:
cols_conte = ['Int', 'TklW', 'Fls', 'Recov', 'Won%']
scores_conte = mids[['Player', 'Pos', 'Squad', 'Age'] + cols_conte]
# Calcular el máximo de cada columna en cols_conte
max_values = scores_conte[cols_conte].max()

# Dividir cada registro por el máximo de su columna correspondiente
scores_conte[cols_conte] = women[cols_conte].div(max_values)

# Sumar las columnas en cols_conte y crear la columna 'suma total'
scores_conte['suma_total'] = scores_conte[cols_conte].sum(axis=1)

# Normalizar la columna 'suma_total' tomando el valor más alto y dividiendo cada registro por ese valor
max_suma_total = scores_conte['suma_total'].max()
scores_conte['score'] = scores_conte['suma_total'] / max_suma_total
display(scores_conte.sort_values('score', ascending=False)[['Player', 'Squad', 'Pos', 'Age', 'score']].head(1))
display(scores_conte[scores_conte['Age'] <= 22].sort_values('score', ascending=False)[['Player', 'Squad','Pos', 'Age', 'score']].head(1))

C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\3227051192.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_conte[cols_conte] = women[cols_conte].div(max_values)
C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\3227051192.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_conte['suma_total'] = scores_conte[cols_conte].sum(axis=1)
C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\3227051192.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

Player     Squad Pos   Age  score
0  Teresa Abelleira  es Spain  MF  23.0    1.0

Player       Squad Pos   Age     score
585  Christy Ucheibe  ng Nigeria  MF  22.0  0.791603

## ej 2: defensa central

In [21]:
defs = women[women['Pos'].str.contains("DF")]
defs = defs[(women['Age'] >= 18) & (women['Age'] < 29)]

C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\1268805327.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  defs = defs[(women['Age'] >= 18) & (women['Age'] < 29)]


In [22]:
cols_def = ['Int', 'TklW', 'Recov', 'Won%']
scores_def = defs[['Player', 'Pos', 'Squad', 'Age'] + cols_def]
# Calcular el máximo de cada columna en cols_def
max_values = scores_def[cols_def].max()

# Dividir cada registro por el máximo de su columna correspondiente
scores_def[cols_def] = women[cols_def].div(max_values)

# Sumar las columnas en cols_def y crear la columna 'suma total'
scores_def['suma_total'] = scores_def[cols_def].sum(axis=1)

# Normalizar la columna 'suma_total' tomando el valor más alto y dividiendo cada registro por ese valor
max_suma_total = scores_def['suma_total'].max()
scores_def['score'] = scores_def['suma_total'] / max_suma_total
display(scores_def.sort_values('score', ascending=False)[['Player', 'Squad', 'Pos', 'Age', 'score']][1:4])
#display(scores_def[scores_def['Age'] <= 22].sort_values('score', ascending=False)[['Player', 'Squad','Pos', 'Age', 'score']].head(1))

C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\1539304551.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_def[cols_def] = women[cols_def].div(max_values)
C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\1539304551.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_def['suma_total'] = scores_def[cols_def].sum(axis=1)
C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\1539304551.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using 

Player           Squad Pos   Age     score
590    Manuela Vanegas     co Colombia  DF  22.0  0.748368
281  Dominique Janssen  nl Netherlands  DF  28.0  0.675566
98     Ellie Carpenter    au Australia  DF  23.0  0.673102

# Varonil

In [23]:
men = pd.read_csv('cop_lib_2023.csv')

In [24]:
men.head()

Rk          Player  Nation    Pos             Squad   Age    Born   90s  \
0   1   Lucas Abascia  ar ARG     DF       cl Ñublense  27.0  1995.0   1.3   
1   2   Miguel Acosta  ar ARG     GK    ar Arg Juniors  21.0  2001.0   1.0   
2   3    Luiz Adriano  br BRA     FW  br Internacional  35.0  1987.0   4.4   
3   4           Adson  br BRA  MF,FW    br Corinthians  22.0  2000.0   3.1   
4   5  Luis Advíncula  pe PER  DF,FW   ar Boca Juniors  32.0  1990.0  12.1   

   CrdY  CrdR  2CrdY  Fls  Fld  Off  Crs  Int  TklW  PKwon  PKcon  OG  Recov  \
0     0     0      0    1    1    0    0    4     3    0.0    0.0   0    8.0   
1     0     0      0    0    0    0    0    0     0    0.0    0.0   0    0.0   
2     0     0      0    3    5    4    0    0     1    0.0    0.0   0    7.0   
3     0     0      0    0    8    0    6    1     4    0.0    0.0   0   17.0   
4     2     0      0    7   17    1   37    5    14    0.0    0.0   0   62.0   

    Won  Lost   Won%  Matches  
0   5.0   2.0   71.4  Matches  
1   1.0   0.0  100.0  Matches  
2   3.0  17.0   15.0  Matches  
3   0.0   5.0    0.0  Matches  
4  13.0  11.0   54.2  Matches

In [25]:
box_2_box = men[men['Pos'].str.contains("MF")]
box_2_box = box_2_box[(men['Age'] >= 18) & (men['Age'] < 29)]

C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\481243247.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  box_2_box = box_2_box[(men['Age'] >= 18) & (men['Age'] < 29)]


### Variables clave para evaluar a un "Box to Box":

    - Fls (Faltas Cometidas):
        Refleja la intensidad defensiva del jugador y su capacidad para cortar jugadas del oponente, aunque es importante que mantenga un equilibrio para evitar sanciones.

    - Fld (Faltas Recibidas):
        Un "Box to Box" tiende a ser un jugador muy involucrado en el juego, tanto en defensa como en ataque, lo que podría llevar a recibir muchas faltas debido a su protagonismo en el mediocampo.

    - Int (Intercepciones):
        Las intercepciones son clave para medir la capacidad del jugador de recuperar el balón en posiciones defensivas y medias, iniciando transiciones rápidas al ataque.

    - TklW (Entradas Ganadas):
        Un jugador "Box to Box" debe ser efectivo en los duelos uno a uno para recuperar balones y cortar el avance del oponente.

    - Recov (Recuperaciones de balón):
        Refleja la capacidad del jugador para recuperar balones en posiciones más adelantadas y en el medio campo, crucial para un centrocampista que cubre mucho terreno.

    - Won% (% de Duelos Ganados):
        Refleja la capacidad del jugador para imponerse en los duelos tanto defensivos como ofensivos, un atributo esencial para un "Box to Box".
    
    - Crs (Centros): Si el jugador tiene funciones ofensivas amplias, podría ser útil evaluar la capacidad de enviar centros al área.

    - Off (Fuera de Juego): Esto puede ser relevante si el jugador tiende a involucrarse mucho en las jugadas ofensivas y llega a posiciones de ataque.

In [26]:
men.columns

Index(['Rk', 'Player', 'Nation', 'Pos', 'Squad', 'Age', 'Born', '90s', 'CrdY',
       'CrdR', '2CrdY', 'Fls', 'Fld', 'Off', 'Crs', 'Int', 'TklW', 'PKwon',
       'PKcon', 'OG', 'Recov', 'Won', 'Lost', 'Won%', 'Matches'],
      dtype='object')

In [31]:
cols_box_2_box = ['Fls', 'Fld', 'Off', 'Crs', 'Int', 'TklW', 'Recov', 'Won%']
scores_b_2_b = box_2_box[['Player', 'Pos', 'Squad', 'Age'] + cols_box_2_box]
# Calcular el máximo de cada columna en cols_box_2_box
max_values = scores_b_2_b[cols_box_2_box].max()

# Dividir cada registro por el máximo de su columna correspondiente
scores_b_2_b[cols_box_2_box] = men[cols_box_2_box].div(max_values)

# Sumar las columnas en cols_box_2_box y crear la columna 'suma total'
scores_b_2_b['suma_total'] = scores_b_2_b[cols_box_2_box].sum(axis=1)

# Normalizar la columna 'suma_total' tomando el valor más alto y dividiendo cada registro por ese valor
max_suma_total = scores_b_2_b['suma_total'].max()
scores_b_2_b['score'] = scores_b_2_b['suma_total'] / max_suma_total
display(scores_b_2_b.sort_values('score', ascending=False)[['Player', 'Squad', 'Pos', 'Age', 'score']][10:20])
#display(scores_b_2_b[scores_b_2_b['Age'] <= 22].sort_values('score', ascending=False)[['Player', 'Squad','Pos', 'Age', 'score']].head(1))

C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\3929703281.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_b_2_b[cols_box_2_box] = men[cols_box_2_box].div(max_values)
C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\3929703281.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  scores_b_2_b['suma_total'] = scores_b_2_b[cols_box_2_box].sum(axis=1)
C:\Users\jonat\AppData\Local\Temp\ipykernel_18560\3929703281.py:14: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a

Player            Squad    Pos   Age     score
419              Lima    br Fluminense  MF,FW  26.0  0.564671
486    Gabriel Menino     br Palmeiras     MF  22.0  0.555554
170    Jesús Castillo  pe Alianza Lima     MF  26.0  0.530942
504  Santiago Montiel   ar Arg Juniors  DF,MF  22.0  0.529304
808  Gabriel Villamíl       bo Bolívar  MF,DF  21.0  0.523158
692     Andrés Romero       ve Monagas  MF,DF  19.0  0.505013
320       Diego Gómez      py Libertad     MF  19.0  0.502382
54              Artur     br Palmeiras  FW,MF  24.0  0.493225
439       Thiago Maia      br Flamengo     MF  25.0  0.491435
784       Alan Varela  ar Boca Juniors     MF  21.0  0.490283

In [39]:
selected = scores_b_2_b[scores_b_2_b['Player'].isin(["Fernando David Cardozo", "Thiago Maia", "Wanderson"])][['Player', 'Squad', 'Pos', 'Age', 'score']]
selected['price'] = ['3M', '6M', '8M']
selected

Player             Squad    Pos   Age     score price
218  Fernando David Cardozo        py Olimpia  MF,FW  21.0  0.830367    3M
439             Thiago Maia       br Flamengo     MF  25.0  0.491435    6M
813               Wanderson  br Internacional  FW,MF  28.0  0.661150    8M

- Fernando David Cardozo	py Olimpia	MF,FW	21.0	0.830367------- 3M
- Thiago Maia	br Flamengo	MF	25.0	0.491435----------------------- 6M
- Wanderson	br Internacional	FW,MF	28.0	0.661150--------------- 8M